### PR TITLE
Support recursive attribute

### PR DESCRIPTION
**Pull request by xuxinyi389**
### PR APIs

```bash

```

### CODE REVIEW

File: paconvert/transformer/basic_transformer.py [hunk 1]
```diff
@@ -191,6 +197,10 @@ def visit_Attribute(self, node):
                 if ".T." in full_attr:
                     # x.T.add
                     is_tensor_api = True
+                if ".data" in full_attr or ".real" in full_attr:
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:200]: - 对于 x.data.add、x.real.add 这种，应该判断  `.data.`、 `.real.`
- 对于 module.weight.add 这种，应该判断 `.weight.` 、`.bias.`

只有data位于中间才必然能对应到三级（`x.data.add`），位于末尾时可三级可二级，判断不精准（ `x.data`）

File: paconvert/transformer/basic_transformer.py [hunk 2]
```diff
@@ -490,6 +502,13 @@ def visit_Call(self, node):
...
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:505]: 从底层转起，遍历到最外层call时，已经变成了：x.weight.detach().zero_()，这里是不是也不用做特殊处理了？

只有 T real 这种不变的需要处理

File: tests/test_Tensor_normal_.py [hunk 3]
```diff
@@ -16,56 +16,65 @@
 
 from apibase import APIBase
 
-obj = APIBase("torch.Tensor.normal_")
+
+class TensorNormal_API(APIBase):
```
Comment (zhwesky2010) [inline tests/test_Tensor_normal_.py:20]: 看一下test_Tensor_uniform_的单测写法

File: tests/test_recursive_attribute.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        Linear = torch.nn.Linear(4, 4)
+        result = Linear.weight
+        """
+    )
+    obj.run(pytorch_code, ["result"])
```
Comment (zhwesky2010) [inline tests/test_recursive_attribute.py:47]: 设置一下 check_value=False 就可以吧

File: paconvert/transformer/basic_transformer.py [hunk 5]
```diff
@@ -75,7 +75,13 @@ def visit_Attribute(self, node):
             super(BasicTransformer, self).generic_visit(node)
 
         # 6.torch.tensor(features_A).T.cuda()
-        if isinstance(node.value, ast.Attribute) and node.value.attr == "T":
+        if isinstance(node.value, ast.Attribute) and node.value.attr in [
+            "T",
+            "real",
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:80]: torch.Tensor.image 也统一做一下吧

File: paconvert/transformer/basic_transformer.py (hunk 1 above)
Comment (xuxinyi389) [reply paconvert/transformer/basic_transformer.py]: 是这样的。当时想着可能会有.T[index].的情形。但是索引的情况工具已经做了很好的适配，此处不需要考虑索引的用法，现在已改正。

File: paconvert/transformer/basic_transformer.py (hunk 2 above)
Comment (xuxinyi389) [reply paconvert/transformer/basic_transformer.py]: .real、.data会被visit_attribute修改为.real()、.detach()。此处导致调用级别超过3的，只有.T. 、.weight.  、.bias. 这类不变的情况需要处理

File: paconvert/transformer/basic_transformer.py (hunk 5 above)
Comment (xuxinyi389) [reply paconvert/transformer/basic_transformer.py:80]: torch.Tensor.imag已添加

File: tests/test_Tensor_normal_.py (hunk 3 above)
Comment (xuxinyi389) [reply tests/test_Tensor_normal_.py]: 已修改

File: paconvert/transformer/basic_transformer.py (hunk 1 above)
Comment (zhwesky2010) [reply paconvert/transformer/basic_transformer.py]: > 是这样的。当时想着可能会有.T[index].的情形。但是索引的情况工具已经做了很好的适配，此处不需要考虑索引的用法，现在已改正。

你这种要把前面看成一个节点，本质上还是二级，`x.data` 这种

File: tests/test_recursive_attribute.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def test_case_6():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        Linear = torch.nn.Linear(4, 4)
+        result = Linear.weight.T.data
+        """
+    )
```
[the remark this answers is not in the document]
Comment (xuxinyi389) [reply tests/test_recursive_attribute.py:252]: 上述所有方式单测均已补全

File: paconvert/transformer/basic_transformer.py [hunk 7]
```diff
@@ -487,9 +502,17 @@ def visit_Call(self, node):
...
-                    # x.T.add
+
+                if (
+                    ".T." in full_attr
+                    or ".weight." in full_attr
+                    or ".bias." in full_attr
+                ):
+                    # x.T.zero_()
+                    # x.weight.zero_()
+                    # x.bias.zero_()
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:513]:
```
x.real.zero_()
x.imag.zero_()
```
的情形有吗

File: tests/test_recursive_attribute.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+# Attribute
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([1.0,2.0])
+        result = x.T.requires_grad
+        """
+    )
+    obj.run(pytorch_code, ["result"], check_value=False)
```
Comment (zhwesky2010) [inline tests/test_recursive_attribute.py:32]: 非随机数的check_value不要设置为False

File: paconvert/transformer/basic_transformer.py (hunk 7 above)
Comment (xuxinyi389) [reply paconvert/transformer/basic_transformer.py:513]: 没有的。x.real.zero_()中，x.real会被visit_attribute转换为x.real()，此处的full_attr为TorchClass.zero_，因此没有上述两种情形。

**Comment by zhwesky2010** [comment]
@xuxinyi389 这个 tests/test_cuda_stream.py::test_case_2 经常容易随机挂，是不是shape太大了还是atol/rtol太高了的原因？要修一下随机挂
